server/src/com/thoughtworks/go/server/service/EnvironmentConfigService.java

```diff
@@ -223,25 +225,157 @@ private void editEnvironments(EditEnvironments edit, EnvironmentConfig config, U
+                    result.setMessage(LocalizedMessage.composite(LocalizedMessage.string("UPDATE_ENVIRONMENT_SUCCESS", environmentName), LocalizedMessage.string("CONFIG_MERGED")));
+                } else if (ConfigSaveState.UPDATED.equals(configSaveState)) {
+                    result.setMessage(LocalizedMessage.string("UPDATE_ENVIRONMENT_SUCCESS", environmentName));
+                }
+            }
+        }, username, result, noPermission, actionFailed);
+    }
+
+    public void deleteEnvironment(final String environmentName, final Username username, final HttpLocalizedOperationResult result) {
+        Localizable noPermission = LocalizedMessage.string("NO_PERMISSION_TO_DELETE_ENVIRONMENT", environmentName, username.getDisplayName());
+        Localizable.CurryableLocalizable actionFailed = LocalizedMessage.string("ENV_DELETE_FAILED", environmentName);
+        editEnvironments(new EditEnvironments() {
+            public void performEdit() {
+                try {
+                    EnvironmentConfig config = getEnvironmentConfig(environmentName);
+                    goConfigService.updateConfig(new DeleteEnvironmentCommand(config, username));
+                    result.setMessage(LocalizedMessage.string("ENVIRONMENT_DELETE_SUCCESSFUL", environmentName));
+                } catch (NoSuchEnvironmentException e) {
+                    result.notFound(LocalizedMessage.string("ENVIRONMENT_NOT_FOUND", environmentName), HealthStateType.general(HealthStateScope.GLOBAL));
+                }
+            }
+        }, username, result, noPermission, actionFailed);
+    }
+
     public interface EditEnvironments {
         void performEdit();
     }
+
+    public static class UpdateEnvironmentCommand implements UpdateConfigCommand, UserAware {
+
+        private final String environmentToUpdate;
+        private final EnvironmentConfig newEnvironmentConfig;
+        private Username username;
+
+        public UpdateEnvironmentCommand(String environmentToUpdate, EnvironmentConfig newEnvironmentConfig, Username username) {
+            this.environmentToUpdate = environmentToUpdate;
+            this.newEnvironmentConfig = newEnvironmentConfig;
+            this.username = username;
+        }
+
+        @Override
+        public CruiseConfig update(CruiseConfig cruiseConfig) throws Exception {
+            EnvironmentsConfig environments = cruiseConfig.getEnvironments();
+            EnvironmentConfig oldConfig = environments.named(new CaseInsensitiveString(environmentToUpdate));
+            int index = environments.indexOf(oldConfig);
+            environments.remove(index);
+            environments.add(index, newEnvironmentConfig);
+            return cruiseConfig;
+        }
+
+        @Override
+        public ConfigModifyingUser user() {
+            return new ConfigModifyingUser(username);
+        }
+    }
+
+    public static class ModifyEnvironmentCommand implements UpdateConfigCommand {
+        private final String uuid;
```

Comment:
Can you think of another name for this class? UpdateEnvironmentCommand and ModifyEnvironmentCommand seem a little confusing :)